### **Prediction** — a block-group crime model that beats the incumbent *inside* cities

The deployed **agency-scale** model (`04_bg_comparison`) levels whole cities well but fits weakly
*within* a city — it can't tell a city's dangerous blocks from its safe ones. This notebook builds a
block-group model on **`log(weighted rate)`** (absolute, no demeaning) and shows it closes that gap.

**Order of the story:**
1. **The incumbent's block-group problem** — strong between cities, weak within, for *both* property
   and total crime (Detroit is the sharpest case).
2. **The refresh on PROPERTY crime (9 cities)** first — the wider sample makes every cross-city number
   less brittle and shows *sample size is the main lever*.
3. **The same refresh on TOTAL crime (5 cities).**
4. **Per-city breakdown, hyperparameter tuning, and next steps.**

Every configuration is scored on the two axes `04` used to indict the incumbent:

| axis | question | metrics |
| --- | --- | --- |
| **LEVEL** (cross-city) | are whole cities placed at the right height? | `level_r`, `r2_oos`, `mae` |
| **RANK** (within-city) | inside a city, are the dangerous blocks flagged? | `recall@top10/30`, `called_safe`, `withinR2` |

- **Splits:** stratified **80/20** (interpolation — a city we've already labeled) and **LOCO**
  (leave-one-city-out — a brand-new city, the honest national-rollout estimate). The gap between them
  is the value of having seen a city before.
- **Predictor ladder:** `ours` → `ours+approved` (de-duplicated incumbent features) →
  `ours+approved+agency` (adds last year's agency-wide crime total). No race / protected predictors.
- **Models:** `RidgeCV` and `LightGBM`, same ladder.


In [1]:
%load_ext autoreload
%autoreload 2
%matplotlib inline

import contextlib
import io
import sys
import warnings
from pathlib import Path

import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
import optuna

_here = Path.cwd().resolve()
_root = next(p for p in [_here, *_here.parents] if (p / "src").is_dir())
sys.path.insert(0, str(_root / "src"))

from crime_blockgroup_mapping.constants import CITIES
from regression_modelling.constants import AGENCY_ANCHOR_COL
from regression_modelling.models import baseline
from regression_modelling.models.cv import load_pooled_table, loco_metrics, within_city_recall
from regression_modelling.models.experiments import (
    ExperimentSpec, run_spec, between_city_level_r, resolve_predictors,
)

pd.set_option("display.width", 240)
pd.set_option("display.max_columns", 40)
warnings.filterwarnings("ignore", category=RuntimeWarning)
warnings.filterwarnings("ignore", message="An input array is constant")
optuna.logging.set_verbosity(optuna.logging.WARNING)

In [2]:
# 5 full-coverage cities carry weighted-total (wtotal); the rest add weighted-property (wprop).
full_cities = [c for c, cfg in CITIES.items() if not cfg.property_only]
all_cities = list(CITIES.keys())

# Pooled block-group tables: one row per BG with predictors, observed rates, and the agency total.
pool5 = load_pooled_table(full_cities)
pool10 = load_pooled_table(all_cities)

# Drop cities whose crime counts are entirely zero. Columbus's upstream bg_crime extract produced
# all-zero counts (BG geometry present, crime never joined) -> a constant rate that makes
# correlation / ranking metrics undefined. Remove it rather than let it poison the LEVEL metrics.
_degenerate = [c for c, g in pool10.groupby("city") if float(g["wprop_rate"].abs().sum()) == 0]
if _degenerate:
    print("dropping all-zero-crime cities (upstream data gap):", _degenerate)
    pool10 = pool10[~pool10["city"].isin(_degenerate)].copy()
    all_cities = [c for c in all_cities if c not in _degenerate]

print("pool5:", pool5.shape, "| pool10:", pool10.shape, "| wprop cities:", len(all_cities))

  houston            1629 BGs
  chicago            2164 BGs
  atlanta             426 BGs
  kansas_city         473 BGs
  detroit             622 BGs
  pooled             5314 BGs across 5 cities
  dropped 22 zero/NaN-pop BGs -> 5292 remain (filtered geoid set for fit + Moran's I + bias join)
  dropped 8 BGs below daytime_pop 100 (small-denominator rate artifacts) -> 5284 remain
  houston            1629 BGs
  chicago            2164 BGs
  atlanta             426 BGs
  sacramento          366 BGs
  san_francisco       679 BGs
  pittsburgh          317 BGs
  kansas_city         473 BGs
  columbus            637 BGs


  jacksonville        550 BGs
  detroit             622 BGs
  pooled             7863 BGs across 10 cities
  dropped 32 zero/NaN-pop BGs -> 7831 remain (filtered geoid set for fit + Moran's I + bias join)
  dropped 12 BGs below daytime_pop 100 (small-denominator rate artifacts) -> 7819 remain
dropping all-zero-crime cities (upstream data gap): ['columbus']
pool5: (5284, 108) | pool10: (7182, 108) | wprop cities: 9


### Benchmark & metric glossary

The `existing (04)*` row is the deployed **agency-scale** model, scored with the **same helpers** as
the refresh (via `baseline.existing_model_run`) so the comparison is apples-to-apples. It is computed
live for **both** targets — wtotal / 5 cities and wprop / 9 cities.

- **`level_r`** — Pearson of city-mean predicted vs observed rate (cross-city LEVEL).
- **`r2_oos` / `mae`** — pooled out-of-sample fit on the absolute log-rate. **The incumbent has none:**
  it ships a within-city *risk ranking* (a points score on a different scale), not a calibrated rate,
  so you can rank with it but not read a number off it. The refresh predicts a calibrated log-rate, so
  it *can* be scored on calibration — that is itself part of the case for the refresh.
- **`recall@top10/30`** — of a city's truly-dangerous blocks (top decile by observed rate), the share
  the model also ranks in its own within-city top 10% / 30% (higher = better).
- **`called_safe`** — share of dangerous blocks buried below the model's within-city median (the costly
  miss; lower = better).
- **`withinR2`** — mean per-city (corr of predicted vs observed log-rate)² (within-city fit).

The incumbent's `level_r` is **in-sample** (it has seen every city); every refresh row is honest
**LOCO** (city fully held out) or optimistic **80/20** — so *beating the incumbent under LOCO is the
strong result*.


In [3]:
_POOL = {"wtotal": pool5, "wprop": pool10}
_RUNS = {}


def get_run(category, pset, split, model="ridge"):
    """Return a cached absolute log-rate run for one (predictor set, split, model).

    Runs are memoised in ``_RUNS`` so each configuration is fit once, and fitter stdout is
    suppressed to keep tables clean. ``pset`` is 'ours' | 'ours+approved' |
    'ours+approved+agency'; ``split`` is 'loco' | '80_20'; ``model`` is 'ridge' | 'gbm'.
    """
    key = (category, pset, split, model)
    if key not in _RUNS:
        spec = ExperimentSpec(pset, "absolute", split, model)
        with contextlib.redirect_stdout(io.StringIO()):
            _RUNS[key] = run_spec(spec, _POOL[category], category)
    return _RUNS[key]


def within_city_r2(scored, category):
    """Mean per-city within-city R2 = (Pearson of y_pred vs log1p(rate))^2.

    Affine-invariant, so it needs no city moments -- the honest 'within-city variance
    explained', comparable to 04's mean per-city R2. Returns the mean across cities.
    """
    rate = f"{category}_rate"
    r2s = []
    # One held-out city at a time: correlate predicted vs observed log-rate within that city.
    for _, g in scored.groupby("holdout_city"):
        if len(g) <= 2:
            continue
        y = np.log1p(g[rate].astype(float))
        p = g["y_pred"].astype(float)
        r = np.corrcoef(y, p)[0, 1]
        if np.isnan(r):          # constant-rate city -> undefined, skip
            continue
        r2s.append(r * r)
    return round(float(np.mean(r2s)), 3) if r2s else np.nan


def incumbent_run(category, cities):
    """Deployed incumbent (agency-scale, 04_bg_comparison) as a refresh-style run dict."""
    with contextlib.redirect_stdout(io.StringIO()):
        return baseline.existing_model_run(category, cities=cities)


def incumbent_row(category, cities):
    """Score the incumbent with the SAME helpers as the refresh -> a benchmark dict.

    The incumbent ships a within-city risk RANKING (points-count score), not a calibrated
    rate, so ``r2_oos`` / ``mae`` are undefined (different scale) and omitted here; only the
    scale-invariant LEVEL / RANK metrics apply.
    """
    run = incumbent_run(category, cities)
    rec = within_city_recall(run, category).loc["POOLED"]
    return {"level_r": between_city_level_r(run),
            "recall@top10": rec["recall@top10"], "recall@top30": rec["recall@top30"],
            "called_safe": rec["called_safe"],
            "withinR2": within_city_r2(run["scored"], category)}


def incumbent_within_by_city(category, cities):
    """Per-city within-city R2 for the incumbent -- exposes where its within-city fit is
    weakest (e.g. Detroit) even though its between-city LEVEL is strong."""
    run = incumbent_run(category, cities)
    rate = f"{category}_rate"
    g = run["scored"].copy()
    g["_obs"] = np.log1p(g[rate].astype(float))
    rows = []
    # One row per city: squared within-city correlation of the incumbent score vs observed rate.
    for c, d in g.groupby("holdout_city"):
        r = np.corrcoef(d["_obs"], d["y_pred"])[0, 1]
        rows.append({"city": c, "n": len(d), "within_r2": round(float(r * r), 3)})
    return pd.DataFrame(rows).set_index("city").sort_values("within_r2")


# Incumbent benchmark rows, computed live for both targets (wtotal / 5, wprop / 9 cities).
EXISTING_WTOTAL = incumbent_row("wtotal", full_cities)
EXISTING_WPROP = incumbent_row("wprop", all_cities)


def combined_table(category, sets, model="ridge", existing=None):
    """Level+rank scorecard for one model across predictor sets, LOCO and 80/20.

    LEVEL columns (whole-city placement): level_r, r2_oos, mae.
    RANK columns (dangerous-block flagging within a city): recall@top10, recall@top30,
    called_safe, withinR2. An optional ``existing`` dict prepends the in-sample incumbent
    (r2_oos / mae blank -- the incumbent ranks, it does not calibrate).
    """
    rows = []
    if existing is not None:
        rows.append({"predictors": "existing (04)*", "split": "in-sample",
                     "level_r": existing["level_r"], "r2_oos": np.nan, "mae": np.nan,
                     "recall@top10": existing["recall@top10"], "recall@top30": existing["recall@top30"],
                     "called_safe": existing["called_safe"], "withinR2": existing["withinR2"]})
    # One block per predictor set; within each, the honest LOCO row then the optimistic 80/20 row.
    for name in sets:
        for split in ["loco", "80_20"]:
            run = get_run(category, name, split, model)
            m = loco_metrics(run).loc["POOLED"]
            rec = within_city_recall(run, category).loc["POOLED"]
            rows.append({"predictors": name, "split": split.replace("_", "/"),
                         "level_r": between_city_level_r(run),
                         "r2_oos": round(float(m["r2_oos"]), 3), "mae": round(float(m["mae"]), 2),
                         "recall@top10": rec["recall@top10"], "recall@top30": rec["recall@top30"],
                         "called_safe": rec["called_safe"],
                         "withinR2": within_city_r2(run["scored"], category)})
    return pd.DataFrame(rows).set_index(["predictors", "split"])


def level_by_city(category, pset, model="ridge", split="loco"):
    """Per-city LEVEL + within-city fit for one config -- decomposes the single `level_r`.

    For each held-out city returns its observed vs predicted city-MEAN log-rate (the city-level
    share the agency total carries) and the within-city correlation (ranking quality). Makes
    visible whether a model's level skill is uniform or carried by 1-2 cities.
    """
    run = get_run(category, pset, split, model)
    rate = f"{category}_rate"
    g = run["scored"].copy()
    g["_obs"] = np.log1p(g[rate].astype(float))
    rows = []
    # One row per held-out city: its mean level (obs vs pred) and within-city rank correlation.
    for city, d in g.groupby("holdout_city"):
        r = np.corrcoef(d["_obs"], d["y_pred"])[0, 1] if len(d) > 2 else np.nan
        rows.append({"city": city, "n": len(d),
                     "obs_mean": round(float(d["_obs"].mean()), 3),
                     "pred_mean": round(float(d["y_pred"].mean()), 3),
                     "level_err": round(float(d["y_pred"].mean() - d["_obs"].mean()), 3),
                     "within_r": round(float(r), 3)})
    out = pd.DataFrame(rows).set_index("city")
    out.attrs["level_r"] = round(float(np.corrcoef(out["obs_mean"], out["pred_mean"])[0, 1]), 3)
    return out

### Predictor sets

Exactly what each rung feeds the model. The ladder is additive: `ours+approved` adds the approved,
de-duplicated incumbent features to `ours`; `ours+approved+agency` adds **one** column,
`agency_lag_<target>_log` — last year's agency-wide crime total, identical across every BG in a city
(so it carries the city-level variance; the BG predictors carry the rest).

In [4]:
# Print the resolved predictor columns for each rung (wtotal pool), showing what each rung adds.
_seen = []
for pset in ["ours", "ours+approved", "ours+approved+agency"]:
    cols = resolve_predictors(ExperimentSpec(pset, "absolute", "loco", "ridge"), "wtotal", pool5)
    added = [c for c in cols if c not in _seen]      # only the columns this rung introduces
    print(f"{pset}  (total {len(cols)}, +{len(added)} new):")
    for c in added:
        print("   +", c)
    _seen = cols
    print()

ours  (total 23, +23 new):
   + moved1yr_pct
   + own_pct
   + lap_pct
   + city_centers_dist
   + pop_est_5mile_log
   + pop_ch_1mile
   + vacant_pct_log
   + clip_liens_pct_log
   + clip_foreclosure_pct_log
   + unq_convenience_stores_clips
   + unq_gas_stations_clips
   + unq_liquor_stores_clips
   + vacant_pct_lag6_log
   + clip_liens_pct_lag6_log
   + clip_foreclosure_pct_lag6_log
   + transit_has_transit
   + transit_service_intensity_logc
   + transit_nearest_stop_m_log
   + transit_overnight_stop_share
   + roof_condition_avg
   + roof_debris_pct_avg
   + hardscapes_pct_avg
   + roof_missing_material_pct

ours+approved  (total 28, +5 new):
   + in_household_pct
   + det_pct
   + div_encentral
   + div_midatlantic
   + div_southatlantic

ours+approved+agency  (total 29, +1 new):
   + agency_lag_wtotal_log



### 1 · The incumbent's block-group problem — property *and* total

The deployed agency-scale model does one thing well and one badly, and it's true on both crime families:

- **Between cities (LEVEL): strong.** It places whole cities at nearly the right height
  (`level_r` ≈ 0.90 total / 0.85 property).
- **Within a city (RANK): weak.** Inside a city it explains little block-to-block variation
  (`withinR2` ≈ 0.32 total / 0.24 property) — and it is weakest exactly where the stakes are highest,
  e.g. **Detroit** (within-R² ≈ 0.12 total / 0.10 property).

That within-city gap is the whole target of the refresh. (Recall the incumbent has no `r2_oos`/`mae`:
it ranks, it doesn't calibrate.)


In [5]:
# The incumbent's own indictment: strong BETWEEN-city LEVEL, weak WITHIN-city fit -- both targets.
print("INCUMBENT (agency-scale) benchmark, scored with the refresh's own helpers\n")
print("  wtotal / 5 cities:", {k: round(v, 3) for k, v in EXISTING_WTOTAL.items()})
print("  wprop  / 9 cities:", {k: round(v, 3) for k, v in EXISTING_WPROP.items()})

# Per-city within-city fit: the incumbent is weakest inside cities, worst of all in Detroit.
print("\nPer-city WITHIN-city fit (R2) -- the gap the refresh targets:")
print("\n  wtotal:")
print(incumbent_within_by_city("wtotal", full_cities).to_string())
print("\n  wprop:")
print(incumbent_within_by_city("wprop", all_cities).to_string())

INCUMBENT (agency-scale) benchmark, scored with the refresh's own helpers

  wtotal / 5 cities: {'level_r': 0.898, 'recall@top10': np.float64(0.285), 'recall@top30': np.float64(0.647), 'called_safe': np.float64(0.163), 'withinR2': 0.319}
  wprop  / 9 cities: {'level_r': 0.853, 'recall@top10': np.float64(0.251), 'recall@top30': np.float64(0.572), 'called_safe': np.float64(0.196), 'withinR2': 0.244}

Per-city WITHIN-city fit (R2) -- the gap the refresh targets:

  wtotal:
                n  within_r2
city                        
detroit       592      0.123
houston      1623      0.250
atlanta       421      0.364
chicago      2154      0.394
kansas_city   464      0.463

  wprop:
                  n  within_r2
city                          
detroit         592      0.099
sacramento      365      0.125
san_francisco   674      0.126
houston        1623      0.171
atlanta         421      0.265
jacksonville    550      0.311
chicago        2154      0.321
pittsburgh      299      0.384
ka

### 2 · The refresh on PROPERTY crime (9 cities)

We lead with **property** because it has the **wider sample** (9 cities vs 5), which makes every
cross-city number less brittle — the clearest way to see that **sample size, not model cleverness, is
the main lever.** Same predictor ladder (`ours → +approved → +agency`), both models, scored on LEVEL
and RANK against the incumbent benchmark from section 1.

Read the two splits as a pair: **80/20** (a city we've already labeled — interpolation) vs **LOCO** (a
brand-new city — extrapolation). Watch the agency lagged total lift cross-city LEVEL, and watch the
80/20→LOCO gap as the price of never having seen a city.


In [6]:
SETS = ["ours", "ours+approved", "ours+approved+agency"]
print(f"CROSS-CITY · absolute log(weighted PROPERTY rate) · wprop / {len(all_cities)} cities · RidgeCV")
print(combined_table("wprop", SETS, model="ridge", existing=EXISTING_WPROP).to_string())
print("\n* existing (04) is IN-SAMPLE; its r2_oos/mae are blank (it ranks, it does not calibrate).")
print("  ours / +approved / +agency rows are honest LOCO and optimistic 80/20.")

CROSS-CITY · absolute log(weighted PROPERTY rate) · wprop / 9 cities · RidgeCV


                                level_r  r2_oos   mae  recall@top10  recall@top30  called_safe  withinR2
predictors           split                                                                              
existing (04)*       in-sample    0.853     NaN   NaN         0.251         0.572        0.196     0.244
ours                 loco         0.550   0.327  0.59         0.339         0.720        0.119     0.325
                     80/20        0.848   0.385  0.57         0.369         0.702        0.128     0.333
ours+approved        loco         0.797   0.358  0.58         0.352         0.728        0.112     0.337
                     80/20        0.957   0.417  0.56         0.411         0.709        0.142     0.355
ours+approved+agency loco         0.964   0.380  0.57         0.357         0.725        0.113     0.336
                     80/20        0.987   0.422  0.55         0.390         0.709        0.142     0.354

* existing (04) is IN-SAMPLE; its r2_oos/mae are blank

In [7]:
print(f"CROSS-CITY · absolute log(weighted PROPERTY rate) · wprop / {len(all_cities)} cities · LightGBM")
print(combined_table("wprop", SETS, model="gbm", existing=EXISTING_WPROP).to_string())
print("\n* same predictor ladder; LightGBM replaces RidgeCV. Compare LOCO rows head-to-head.")

CROSS-CITY · absolute log(weighted PROPERTY rate) · wprop / 9 cities · LightGBM


                                level_r  r2_oos   mae  recall@top10  recall@top30  called_safe  withinR2
predictors           split                                                                              
existing (04)*       in-sample    0.853     NaN   NaN         0.251         0.572        0.196     0.244
ours                 loco         0.291   0.270  0.63         0.342         0.701        0.117     0.349
                     80/20        0.965   0.460  0.54         0.348         0.723        0.135     0.402
ours+approved        loco         0.659   0.360  0.59         0.349         0.709        0.114     0.365
                     80/20        0.990   0.483  0.53         0.397         0.738        0.099     0.428
ours+approved+agency loco         0.789   0.359  0.59         0.367         0.709        0.123     0.358
                     80/20        0.991   0.490  0.52         0.390         0.730        0.106     0.432

* same predictor ladder; LightGBM replaces RidgeCV. Co

**Read — property refresh (9 cities).**

- **We beat the incumbent on RANK — the block-level job it fails — under honest LOCO.** Both models:
  `withinR2` ≈ 0.33–0.43 vs incumbent **0.244**, `recall@top30` ≈ 0.70–0.74 vs **0.572**,
  `recall@top10` ≈ 0.34–0.41 vs **0.251**, `called_safe` ≈ 0.11–0.14 vs **0.196**. The refresh tells a
  city's dangerous blocks apart; the incumbent barely does.
- **The agency lagged total is a decisive LEVEL predictor, used at full strength here.** RidgeCV LOCO
  `level_r` climbs monotonically **0.55 → 0.80 → 0.96** (`r2_oos` 0.33 → 0.36 → 0.38). At the top rung
  Ridge's *honest LOCO* level (**0.964**) actually exceeds the incumbent's *in-sample* 0.853 — we place
  unseen cities better than the deployed model places cities it has seen.
- **Sample size is the main lever, not model cleverness.** Every 80/20 number dominates its LOCO twin
  (e.g. `ours` Ridge `level_r` 0.85 vs 0.55) — the gap is purely "have we labeled this city." That gap,
  not the Ridge-vs-GBM choice, is the biggest swing on the page.
- **Ridge vs GBM here: Ridge wins LEVEL (0.96 vs 0.79), RANK is a near-tie** (GBM `withinR2` 0.358 vs
  Ridge 0.336; `recall@top10` 0.367 vs 0.357). On this wider target the linear model + agency predictor
  is the cleaner all-round performer.


### 3 · The same refresh on TOTAL crime (5 cities)

Weighted-**total** covers only the 5 full-coverage cities, so its cross-city `level_r` is a 5-point
correlation — powerful but fragile (the per-city breakdown below opens it up). Same ladder, both
models, against the wtotal incumbent benchmark.


In [8]:
print("CROSS-CITY · absolute log(weighted rate) · wtotal / 5 full-coverage cities · RidgeCV")
print(combined_table("wtotal", SETS, model="ridge", existing=EXISTING_WTOTAL).to_string())
print("\n* existing (04) is IN-SAMPLE; r2_oos/mae blank. Refresh rows are honest LOCO and 80/20.")

CROSS-CITY · absolute log(weighted rate) · wtotal / 5 full-coverage cities · RidgeCV


                                level_r  r2_oos   mae  recall@top10  recall@top30  called_safe  withinR2
predictors           split                                                                              
existing (04)*       in-sample    0.898     NaN   NaN         0.285         0.647        0.163     0.319
ours                 loco         0.139   0.245  0.72         0.421         0.788        0.077     0.422
                     80/20        0.902   0.502  0.60         0.304         0.775        0.049     0.447
ours+approved        loco        -0.023   0.232  0.72         0.409         0.782        0.079     0.428
                     80/20        0.957   0.514  0.59         0.314         0.775        0.069     0.448
ours+approved+agency loco         0.194   0.194  0.75         0.411         0.784        0.085     0.426
                     80/20        0.963   0.515  0.59         0.324         0.765        0.069     0.449

* existing (04) is IN-SAMPLE; r2_oos/mae blank. Refres

In [9]:
print("CROSS-CITY · absolute log(weighted rate) · wtotal / 5 cities · LightGBM")
print(combined_table("wtotal", SETS, model="gbm", existing=EXISTING_WTOTAL).to_string())
print("\n* same predictor ladder as the RidgeCV table; compare LOCO rows head-to-head.")

CROSS-CITY · absolute log(weighted rate) · wtotal / 5 cities · LightGBM


                                level_r  r2_oos   mae  recall@top10  recall@top30  called_safe  withinR2
predictors           split                                                                              
existing (04)*       in-sample    0.898     NaN   NaN         0.285         0.647        0.163     0.319
ours                 loco         0.592   0.353  0.70         0.363         0.758        0.109     0.380
                     80/20        0.952   0.516  0.60         0.333         0.735        0.039     0.461
ours+approved        loco         0.569   0.379  0.68         0.393         0.772        0.095     0.404
                     80/20        0.948   0.521  0.60         0.392         0.745        0.049     0.472
ours+approved+agency loco         0.691   0.375  0.69         0.397         0.758        0.101     0.398
                     80/20        0.966   0.529  0.60         0.363         0.745        0.069     0.475

* same predictor ladder as the RidgeCV table; compare 

**Read — total refresh (5 cities).**

- **Same RANK win over the incumbent, if anything larger.** RidgeCV LOCO `withinR2` ≈ 0.43 vs incumbent
  **0.319**, `recall@top10` ≈ 0.41 vs **0.285**, `recall@top30` ≈ 0.78 vs **0.647**, `called_safe`
  ≈ 0.08 vs **0.163** (roughly half the costly misses). This is the Detroit-style within-city gap being
  closed.
- **But cross-city LEVEL is fragile on 5 cities.** RidgeCV LOCO `level_r` is 0.14 / −0.02 / 0.19 across
  the ladder — near-zero and unstable — while 80/20 sits at 0.90–0.96. GBM holds LEVEL far better under
  LOCO (0.59 / 0.57 / **0.69**).
- **Note the flip vs property.** On wtotal/5, GBM wins LEVEL (0.69 vs Ridge 0.19); on wprop/9, Ridge
  wins LEVEL (0.96 vs 0.79). A verdict that flips with the target is a small-sample artifact — the
  per-city breakdown shows it is one city (Atlanta) driving wtotal LEVEL, not a real skill gap.
- **Both models beat the incumbent on RANK under LOCO on both targets** — the robust, repeatable result.


### 4 · Per-city breakdown — is `level_r` uniform or carried by 1–2 cities?

`level_r` is one correlation over 5 (wtotal) or 9 (wprop) city-mean points, so a single city can swing
it. `level_by_city` opens it up per held-out city:
- **`level_err`** = predicted mean − observed mean (LOCO). Near 0 = the model places the city at the
  right height; a large value that flips sign across cities is what makes `level_r` noisy.
- **`within_r`** = within-city correlation (ranking quality), the offset-free product axis.

Watch whether each model's LEVEL skill is uniform or rides on a couple of cities — and note how steady
`within_r` (the RANK product) stays regardless.


In [10]:
# Compare the two models' per-city level + within-city fit on the top rung (ours+approved+agency).
CFG = "ours+approved+agency"
for cat in ["wtotal", "wprop"]:
    # One target at a time; print Ridge then GBM so the level_err / within_r columns sit side by side.
    for model in ["ridge", "gbm"]:
        tbl = level_by_city(cat, CFG, model=model)
        print(f"== {cat} · {CFG} · {model} · LOCO (between-city level_r = {tbl.attrs['level_r']}) ==")
        print(tbl.to_string())
        print()

== wtotal · ours+approved+agency · ridge · LOCO (between-city level_r = 0.194) ==
                n  obs_mean  pred_mean  level_err  within_r
city                                                       
atlanta       369     3.414      5.279      1.865     0.715
chicago      2064     3.443      3.100     -0.344     0.690
detroit       600     4.201      4.548      0.347     0.457
houston      1536     3.446      3.594      0.148     0.593
kansas_city   445     3.742      3.566     -0.176     0.763

== wtotal · ours+approved+agency · gbm · LOCO (between-city level_r = 0.691) ==
                n  obs_mean  pred_mean  level_err  within_r
city                                                       
atlanta       369     3.414      3.987      0.573     0.661
chicago      2064     3.443      3.593      0.150     0.666
detroit       600     4.201      4.488      0.287     0.471
houston      1536     3.446      3.283     -0.163     0.576
kansas_city   445     3.742      3.441     -0.302     0.7

**Read — per-city breakdown (why `level_r` moves).** `level_r` is a correlation over a handful of city
means, so one miscalibrated city dominates it.

- **wtotal / 5 — Ridge is sunk by a single city.** Atlanta `level_err` **+1.865** (predicts 5.28 vs
  observed 3.41) drags Ridge's `level_r` to **0.194** while the other four sit within ±0.35. GBM has no
  blow-up (worst |err| 0.573) → `level_r` **0.691**. This lone outlier — not a general skill gap — is
  the whole wtotal LEVEL story. Do **not** tune toward `level_r` on n = 5.
- **wprop / 9 — with the agency total, Ridge is well-calibrated everywhere.** Every city's `level_err`
  is within ±0.17 → `level_r` **0.964**. GBM instead over-predicts Atlanta (**+0.535**) and
  under-predicts Kansas City (−0.316) → `level_r` **0.789**. The linear model's even calibration is what
  wins LEVEL on the wider target.
- **`within_r` is healthy and stable for both** (≈ 0.37–0.77 per city, no negatives) — the RANK product
  holds up city-by-city, independent of the LEVEL wobble. Judge extrapolation on the 9-city property
  evidence, not wtotal/5.


### 5 · Hyperparameter tuning — does Optuna move LightGBM?

The GBM above uses fixed defaults. Here we let **Optuna** search LightGBM's hyperparameters, then ask
whether it changes the verdict. Two guardrails against fooling ourselves on 9 cities:
- **Objective = 80/20 `r2_oos` (interpolation), not LOCO.** With only 9 cities, tuning *toward* the
  LOCO `level_r` would chase a 9-point correlation and overfit the fold structure. We tune on the
  stable within-city split and then read the **LOCO** numbers *untouched* as the honest check.
- **We judge impact on LOCO**, the number that reflects the national rollout.


In [11]:
# Tune LightGBM on the PROPERTY target (9 cities), the wider/less-brittle sample.
TUNE_CAT, TUNE_PSET = "wprop", "ours+approved+agency"


def _gbm_pooled(params, split, field):
    """Fit tuned-GBM through the same run_spec pipeline; return one POOLED LEVEL metric."""
    spec = ExperimentSpec(TUNE_PSET, "absolute", split, "gbm", gbm_params=params)
    with contextlib.redirect_stdout(io.StringIO()):
        run = run_spec(spec, pool10, TUNE_CAT)
    return float(loco_metrics(run).loc["POOLED"][field])


def _objective(trial):
    """Maximise 80/20 (interpolation) r2_oos -- a stable within-city objective, NOT LOCO level_r."""
    params = dict(
        n_estimators=trial.suggest_int("n_estimators", 200, 1200, step=100),
        learning_rate=trial.suggest_float("learning_rate", 0.01, 0.1, log=True),
        num_leaves=trial.suggest_int("num_leaves", 8, 127),
        min_child_samples=trial.suggest_int("min_child_samples", 10, 150),
        subsample=trial.suggest_float("subsample", 0.6, 1.0),
        colsample_bytree=trial.suggest_float("colsample_bytree", 0.6, 1.0),
        reg_alpha=trial.suggest_float("reg_alpha", 0.0, 5.0),
        reg_lambda=trial.suggest_float("reg_lambda", 0.0, 5.0),
    )
    return _gbm_pooled(params, "80_20", "r2_oos")


study = optuna.create_study(direction="maximize", sampler=optuna.samplers.TPESampler(seed=42))
study.optimize(_objective, n_trials=40, show_progress_bar=False)
best = study.best_params
print("best 80/20 r2_oos:", round(study.best_value, 4))
print("best params:", {k: (round(v, 4) if isinstance(v, float) else v) for k, v in best.items()})


def _scorecard(params, tag):
    """Print LOCO + 80/20 LEVEL/RANK for a param set through the same pipeline."""
    for split in ["loco", "80_20"]:
        spec = ExperimentSpec(TUNE_PSET, "absolute", split, "gbm", gbm_params=params)
        with contextlib.redirect_stdout(io.StringIO()):
            run = run_spec(spec, pool10, TUNE_CAT)
        m = loco_metrics(run).loc["POOLED"]
        rec = within_city_recall(run, TUNE_CAT).loc["POOLED"]
        print(f"  {tag:8s} {split:5s} | level_r={between_city_level_r(run):.3f} "
              f"r2_oos={m['r2_oos']:.3f} mae={m['mae']:.3f} "
              f"rec@10={rec['recall@top10']:.3f} rec@30={rec['recall@top30']:.3f}")


print("\nwprop · default vs Optuna-tuned LightGBM (LOCO read is untouched by the objective):")
_scorecard(None, "default")
_scorecard(best, "tuned")

best 80/20 r2_oos: 0.501
best params: {'n_estimators': 700, 'learning_rate': 0.0272, 'num_leaves': 21, 'min_child_samples': 14, 'subsample': 0.6931, 'colsample_bytree': 0.6919, 'reg_alpha': 1.7202, 'reg_lambda': 1.7319}

wprop · default vs Optuna-tuned LightGBM (LOCO read is untouched by the objective):


  default  loco  | level_r=0.789 r2_oos=0.359 mae=0.590 rec@10=0.367 rec@30=0.709


  default  80_20 | level_r=0.991 r2_oos=0.490 mae=0.520 rec@10=0.390 rec@30=0.730


  tuned    loco  | level_r=0.799 r2_oos=0.380 mae=0.580 rec@10=0.372 rec@30=0.716


  tuned    80_20 | level_r=0.995 r2_oos=0.501 mae=0.520 rec@10=0.433 rec@30=0.745


**Read — tuning.** Optuna moves the objective it was given (80/20 `r2_oos` 0.490 → **0.501**), and —
the honest test — the **untouched LOCO** numbers improve too but only modestly:

- LOCO `r2_oos` **0.359 → 0.380** (+0.021 calibration), `mae` 0.590 → 0.580, `recall@top10`
  0.367 → 0.372, `recall@top30` 0.709 → 0.716, `level_r` 0.789 → 0.799. No metric goes backwards.
- **Verdict: a real but second-order lever.** Tuning is worth keeping (best params: shallower trees,
  `num_leaves` 21, strong `min_child_samples`/L1-L2 regularization — i.e. Optuna asked for *less*
  overfitting), but the deltas are small next to the two first-order levers: adding the **agency
  predictor** and adding **cities**. We deliberately tuned on 80/20 and read LOCO untouched, so this is
  not a number we optimized into looking good.


### Next steps

1. **More cities — the highest-leverage move.** Every cross-city number here rests on 5–9 cities, and
   LOCO is harsh precisely because each held-out city is structurally unique. Adding POC cities shrinks
   the 80/20→LOCO gap and de-fragilizes `level_r` far more than any modelling change.
2. **Tuning (section 5).** Optuna buys a modest, honest lift on LOCO calibration — worth keeping, but
   it is a second-order lever behind sample size and features. Do **not** tune toward `level_r` on this
   few cities.
3. **Predictors — target *within-city* signal.** The agency lag already carries the between-city level;
   new features should explain block-to-block variation (the incumbent's weak spot and Detroit's):
   finer POI mix, local built-environment / imagery, temporal crime lags where available, and
   mobility / foot-traffic. Brainstorm and screen these the same way — LOCO first, RANK metrics first.
